backend/cfg_selectgen.ml

```diff
@@ -742,43 +753,75 @@ class virtual selector_generic =
             Misc.fatal_error
               "Selection.emit_expr: Return with too many arguments"))
 
-    method emit_expr_aux_trywith env bound_name e1 exn_cont v e2
+    method emit_expr_aux_trywith env bound_name e1 exn_cont v ~extra_args e2
         (_dbg : Debuginfo.t) (_value_kind : Cmm.kind_for_unboxing) =
       (* CR-someday xclerc for xclerc: use the `_dbg` parameter *)
       assert (Sub_cfg.exit_has_never_terminator sub_cfg);
       let exn_label = Cmm.new_label () in
       let env_body = Select_utils.env_enter_trywith env exn_cont exn_label in
+      (* For each exception handler having extra arguments, distinguished
+         registers corresponding to such arguments are created. They are
+         populated at each raise site, which is the only place Cmm allows extra
+         arguments to be involved in a raise. Any function call that involves
+         extra arguments on its exception continuation has to compiled using a
+         wrapper; see [To_cmm_expr.translate_apply]. *)
+      let extra_arg_regs_split =
+        List.map (fun (_param, machtype) -> self#regs_for machtype) extra_args
+      in
+      let extra_arg_regs = Array.concat extra_arg_regs_split in
+      let env_body =
+        env_add_regs_for_exception_extra_args exn_cont extra_arg_regs env_body
+      in
       let r1, s1 = self#emit_sequence env_body e1 ~bound_name in
-      let rv = self#regs_for typ_val in
+      let rv_list =
+        List.map
+          (fun machtype -> self#regs_for machtype)
+          (typ_val :: List.map snd extra_args)
+      in
```

Comment:
I think that we could use `let rv_list = (self#regs_for typ_val) :: extra_args_regs_split`, and remove the extra moves at the beginning of the handler (only keeping the move from `Proc.loc_exn_bucket` to `List.hd rv_list`.
What this PR currently does is reasonable, creating one set of registers for the calling convention of the handler and one set for the arguments of the handler, but since they are both local to the current function I think it's going to be simpler to just use the same set for both.